app/Http/Controllers/AccountController.php

```diff
@@ -169,20 +171,32 @@ public function updateEmail()
 
     public function updateNotificationOptions()
     {
-        $request = request();
+        $requestParams = request()->all()['user_notification_option'] ?? [];
+        if (!is_array($requestParams)) {
+            abort(422);
+        }
 
-        $name = $request['name'] ?? null;
-        $params = get_params($request, 'user_notification_option', ['details:any']);
+        return DB::transaction(function () use ($requestParams) {
+            // FIXME: less queries
+            foreach ($requestParams as $key => $value) {
+                if (!UserNotificationOption::supportsNotifications($key)) {
+                    continue;
+                }
 
-        $option = auth()->user()->notificationOptions()->firstOrCreate(['name' => $name]);
+                $params = get_params($value, null, ['details:any']);
 
-        if ($option->update($params)) {
-            return response(null, 204);
-        } else {
-            return response(['form_error' => [
-                'user_notification_option' => $option->validationErrors()->all(),
-            ]]);
-        }
+                $option = auth()->user()->notificationOptions()->firstOrCreate(['name' => $key]);
+                if (!$option->update($params)) {
```

Comment:
looking again, I probably should've just written this but while there's no `updateOrExplode`, there's `->fill($params)->saveOrExplode()`.